File: Matlab/CCD_2.m
%%  2D Cyclic Coordinate Descent Algorithm
%   Avery Bodenstein
%   9/1/2018

%%  Inputs
clc; clear; close all;

target = [1;0.5]; % target point

num_of_link = 7; % number of links
%   Set up initial joint X locations
xdata = (0:num_of_link);
%   Set up initial joint Y locations
ydata = zeros(1,num_of_link+1);
%   Set up initial joint angles
angledata = zeros(1,num_of_link+1);
%   Error threashold
threashold = 0.5;

%   Set up figure
figure
title('Planar Robot')
axis([-10 10 -10 10])
hold on

%% Run CCD Algorithm

hnd = plot(xdata,ydata,'-r');    % plot robot manipulator
plot(target(1),target(2),'*')    % plot target

input('Enter to Continue...');

[output_angles, output_positions] = CCD_2D(angledata,[xdata;ydata],0,target,threashold,10);

for i = 1:1:size(xdata,2)
    plot(output_positions(1,:), output_positions(2,:), 'ob')
end

set(hnd,'xdata',output_positions(1,:),'ydata',output_positions(2,:)); % update data to plot  

File: Python/CCD_2.py
import numpy as np
import matplotlib.pyplot as plt
from CCD_2D import CCD_2D

target = np.array([1,0.5])
n_link = 7

x_data = np.arange(n_link+1)
y_data = np.zeros( (n_link+1,1) )
angle_data = np.zeros((n_link+1,1))
threashold = 0.5

plt.figure(1, figsize=(9,3))
plt.plot(x_data,y_data,'r-')
plt.plot(target[0],target[1],'b*')
plt.show()

# Run CCD Algorithm

output = CCD_2D(x_data,y_data,angle_data,0,target,threashold,10)
new_x_data = output[0]
new_y_data = output[1]
new_angle_data = output[2]

plt.figure(1, figsize=(9,3))
plt.plot(new_x_data,new_y_data,'r-')
plt.plot(target[0],target[1],'b*')
for i in range(0,np.size(x_data)):
    plt.plot(x_data[i],y_data[i],'ob')
plt.show()


